src/dataflow/src/sink/kafka.rs

```diff
@@ -253,18 +258,79 @@ impl Drop for KafkaSinkToken {
     }
 }
 
+#[derive(Clone)]
+struct KafkaTxProducer {
+    inner: Arc<ThreadedProducer<SinkProducerContext>>,
+    timeout: Duration,
+}
+
+impl KafkaTxProducer {
+    async fn init_transactions(&self) -> KafkaResult<()> {
+        let self_producer = Arc::clone(&self.inner);
+        let self_timeout = self.timeout;
+        task::spawn_blocking(move || self_producer.init_transactions(self_timeout))
+            .await
+            .unwrap_or(Err(KafkaError::Canceled))
+    }
+
+    async fn begin_transaction(&self) -> KafkaResult<()> {
+        let self_producer = Arc::clone(&self.inner);
+        task::spawn_blocking(move || self_producer.begin_transaction())
+            .await
+            .unwrap_or(Err(KafkaError::Canceled))
+    }
+
+    async fn commit_transaction(&self) -> KafkaResult<()> {
+        let self_producer = Arc::clone(&self.inner);
+        let self_timeout = self.timeout;
+        task::spawn_blocking(move || self_producer.commit_transaction(self_timeout))
+            .await
+            .unwrap_or(Err(KafkaError::Canceled))
+    }
+
+    async fn abort_transaction(&self) -> KafkaResult<()> {
+        let self_producer = Arc::clone(&self.inner);
+        let self_timeout = self.timeout;
+        task::spawn_blocking(move || self_producer.abort_transaction(self_timeout))
+            .await
+            .unwrap_or(Err(KafkaError::Canceled))
+    }
+
+    async fn flush(&self) -> KafkaResult<()> {
+        let self_producer = Arc::clone(&self.inner);
+        let self_timeout = self.timeout;
+        task::spawn_blocking(move || self_producer.flush(self_timeout))
+            .await
+            .map_err(|_| KafkaError::Canceled)
```

Comment:
NIT: since you're not actually mapping the error `.or(Err(KafkaError::Cancelled))` looks more idiomatic